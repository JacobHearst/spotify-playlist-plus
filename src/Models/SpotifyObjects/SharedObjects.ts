export interface ImageObject {
    height: number
    width: number
    url: string
}

export interface PublicUserObject {
    display_name?: string
    href: string
    id: string
    images: ImageObject[]
}