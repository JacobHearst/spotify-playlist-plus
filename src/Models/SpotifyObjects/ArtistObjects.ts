/**
 * Response objects that match the Spotify Artist API
 */

import { ImageObject } from "./SharedObjects";

export interface ArtistObject {
    genres: string[]
    href: string
    id: string
    images: ImageObject[]
    name: string
    popularity: number
    uri: string
}

export interface SimplifiedArtistObject {
    href: string
    id: string
    name: string
}