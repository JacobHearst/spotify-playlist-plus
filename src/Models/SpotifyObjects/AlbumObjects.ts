import { ArtistObject, SimplifiedArtistObject } from "./ArtistObjects";
import { ImageObject } from "./SharedObjects";

enum AlbumType {
    Album="album",
    Single="single",
    Compilation="compilation"
}

enum AlbumGroup {
    Album="album",
    Single="single",
    Compilation="compilation",
    AppearsOn="appears_on",
}

/**
 * Response objects that match the Spotify Albums API
 */

export interface AlbumObject {
    album_type: AlbumType
    artists: ArtistObject[]
    genres: string[]
    href: string
    id: string
    images: ImageObject[]
    label: string
    name: string
    popularity: number
    release_date: string
}

export interface SimplifiedAlbumObject {
    album_group: AlbumGroup
    album_type: AlbumType
    artists: SimplifiedArtistObject[]
    id: string
    images: ImageObject[]
    name: string
    release_date: string

}