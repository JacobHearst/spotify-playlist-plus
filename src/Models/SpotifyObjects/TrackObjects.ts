import { SimplifiedAlbumObject } from "./AlbumObjects";
import { ArtistObject } from "./ArtistObjects";

/**
 * Response objects that match the Spotify Tracks API
 */

export interface TrackObject {
    album?: SimplifiedAlbumObject
    artists: ArtistObject[]
    duration_ms: number
    explicit: boolean
    href: string
    id: string
    is_playable: boolean
    name: string
    popularity: number
    track_number: number
}