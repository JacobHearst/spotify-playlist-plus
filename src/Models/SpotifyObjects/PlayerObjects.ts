// from https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/spotify-web-playback-sdk/index.d.ts
// this isn't plagarism right.....?
import { TrackObject as Track } from "./TrackObjects"

declare global {
    interface Window {
        onSpotifyWebPlaybackSDKReady(): void
        Spotify: typeof Spotify
    }
}

/* eslint-disable */
export declare namespace Spotify {
    const Player: typeof SpotifyPlayer;

    export interface Error {
        message: string;
    }

    export type ErrorTypes = 'account_error' | 'authentication_error' | 'initialization_error' | 'playback_error';

    interface PlaybackContext {
        metadata: any;
        uri: string | null;
    }

    interface PlaybackDisallows {
        pausing: boolean;
        peeking_next: boolean;
        peeking_prev: boolean;
        resuming: boolean;
        seeking: boolean;
        skipping_next: boolean;
        skipping_prev: boolean;
    }

    interface PlaybackRestrictions {
        disallow_pausing_reasons: string[];
        disallow_peeking_next_reasons: string[];
        disallow_peeking_prev_reasons: string[];
        disallow_resuming_reasons: string[];
        disallow_seeking_reasons: string[];
        disallow_skipping_next_reasons: string[];
        disallow_skipping_prev_reasons: string[];
    }

    interface PlaybackState {
        context: PlaybackContext;
        disallows: PlaybackDisallows;
        duration: number;
        paused: boolean;
        position: number;
        /**
         * 0: NO_REPEAT
         * 1: ONCE_REPEAT
         * 2: FULL_REPEAT
         */
        repeat_mode: 0 | 1 | 2;
        shuffle: boolean;
        restrictions: PlaybackRestrictions;
        track_window: PlaybackTrackWindow;
    }

    interface PlaybackTrackWindow {
        current_track: Track;
        previous_tracks: Track[];
        next_tracks: Track[];
    }

    interface PlayerInit {
        name: string;
        getOAuthToken(cb: (token: string) => void): void;
        volume?: number;
    }

    type ErrorListener = (err: Error) => void;
    type PlaybackInstanceListener = (inst: WebPlaybackInstance) => void;
    type PlaybackStateListener = (s: PlaybackState) => void;

    type AddListenerFn =
        & ((event: 'ready' | 'not_ready', cb: PlaybackInstanceListener) => void)
        & ((event: 'player_state_changed', cb: PlaybackStateListener) => void)
        & ((event: ErrorTypes, cb: ErrorListener) => void);

    class SpotifyPlayer {
        readonly _options: PlayerInit & {id: string};
        constructor(options: PlayerInit);

        connect(): Promise<boolean>;
        disconnect(): void;
        getCurrentState(): Promise<PlaybackState | null>;
        getVolume(): Promise<number>;
        nextTrack(): Promise<void>;

        addListener: AddListenerFn;
        on: AddListenerFn;

        removeListener(
            event: 'ready' | 'not_ready' | 'player_state_changed' | ErrorTypes,
            cb?: ErrorListener | PlaybackInstanceListener | PlaybackStateListener,
        ): void;

        pause(): Promise<void>;
        previousTrack(): Promise<void>;
        resume(): Promise<void>;
        seek(pos_ms: number): Promise<void>;
        setName(name: string): Promise<void>;
        setVolume(volume: number): Promise<void>;
        togglePlay(): Promise<void>;
    }

    interface WebPlaybackInstance {
        device_id: string;
    }
}
/* eslint-enable */