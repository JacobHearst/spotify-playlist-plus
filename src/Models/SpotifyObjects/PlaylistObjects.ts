import { Tracing } from "trace_events";
import { ImageObject, PublicUserObject } from "./SharedObjects";
import { TrackObject } from "./TrackObjects";

/**
 * Response objects that match the Spotify Playlist API
 */

export interface PlaylistObject {
    description: string
    href: string
    id: string
    images: ImageObject[]
    name: string
    owner: PublicUserObject
    public: boolean
    tracks: PlaylistTrackObject[]
}

export interface PlaylistTrackObject extends TrackObject {
    added_at: string
    track: TrackObject
}