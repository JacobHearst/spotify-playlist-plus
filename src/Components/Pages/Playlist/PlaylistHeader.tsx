import React from "react"
import { Button, ButtonGroup, Col, Dropdown, Image, Row } from "react-bootstrap"
import PlayerEndpoints from "../../../Endpoints/Player"
import TrackService from "../../../Services/TrackService"
import { CaretDownFill } from "react-bootstrap-icons"
import { AuthenticationContext } from "../../../Models/Authentication"
import { PlaylistObject } from "../../../Models/SpotifyObjects/PlaylistObjects"
import { msToSentence } from "../../../Services/Utility"

interface PlaylistHeaderProps {
    playlist?: PlaylistObject
}

interface PlaylistHeaderState extends PlaylistHeaderProps {}

export default class PlaylistHeader extends React.Component<PlaylistHeaderProps, PlaylistHeaderState> {
    static contextType = AuthenticationContext

    constructor(props: PlaylistHeaderProps) {
        super(props)
        this.state = {
            ...props,
        }

        this.playByIntensity = this.playByIntensity.bind(this)
    }

    playByIntensity(decreasing: boolean = false) {
        if (this.state.playlist && this.context.player) {
            const deviceId = this.context.player._options.id
            const tracks = this.state.playlist.tracks.map(({ track }) => track)
            TrackService.intensitySort(tracks, decreasing).then((tracks) => {
                const uris = tracks.map(({ uri }) => uri)
                PlayerEndpoints.startResume(deviceId, uris)
            })
        }
    }

    render() {
        if (!this.props.playlist) {
            return <p>Loading</p>
        }

        const { name, description, owner, tracks } = this.props.playlist!

        let playlistLength = 0
        tracks.forEach(({ track }) => (playlistLength += track.duration_ms))

        return (
            <Row style={{ marginBottom: 15 }}>
                {/* Header */}
                <Col xs="auto">
                    {/* Image size can be up to 640 by 640 */}
                    <Image src={this.props.playlist!.images[0].url} width={250} height={250} style={{ border: "1px solid black" }} />
                </Col>
                <Col>
                    <h2>{name}</h2>
                    <p>{description}</p>
                    <p>Owned by: {owner.display_name}</p>
                    <p>
                        {tracks.length} Songs. {msToSentence(playlistLength)}
                    </p>
                    <Dropdown as={ButtonGroup}>
                        <Button variant="success" disabled>
                            Play
                        </Button>

                        <Dropdown.Toggle split variant="success">
                            <CaretDownFill />
                        </Dropdown.Toggle>

                        <Dropdown.Menu>
                            {/* Is this not just the same as sorting the playlist by intensity and playing it? Shuffle doesn't seem like the right word */}
                            <Dropdown.Item onClick={() => this.playByIntensity()}>Play by intensity (increasing)</Dropdown.Item>
                            <Dropdown.Item onClick={() => this.playByIntensity(true)}>Play by intensity (decreasing)</Dropdown.Item>
                        </Dropdown.Menu>
                    </Dropdown>
                </Col>
            </Row>
        )
    }
}
