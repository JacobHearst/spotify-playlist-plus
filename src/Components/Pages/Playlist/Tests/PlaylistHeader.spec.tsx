import React from 'react'
import renderer from 'react-test-renderer'
import { PlaylistObject } from '../../../../Models/SpotifyObjects/PlaylistObjects'
import { PublicUserObject } from '../../../../Models/SpotifyObjects/SharedObjects'
import PlaylistHeader from '../PlaylistHeader'

describe("PlaylistHeader", () => {
    const mockPlaylist: PlaylistObject = {
        description: "My description",
        id: "myPlaylistId",
        href: "https://spotify.com",
        images: [{ url: "https://spotify.com", height: 100, width: 100 }],
        name: "My Playlist",
        owner: { display_name: "My user" } as PublicUserObject,
        public: true,
        tracks: []
    }

    it("should properly display playlist information when a playlist is passed in", () => {
        const component = renderer.create(<PlaylistHeader playlist={mockPlaylist} />)

        let tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })

    it("should display a loading message when no playlist is passed in", () => {
        const component = renderer.create(<PlaylistHeader />)
        
        let tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })
})
