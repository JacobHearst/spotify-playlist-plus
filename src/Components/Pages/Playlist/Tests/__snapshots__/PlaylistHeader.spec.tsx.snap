// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaylistHeader should display a loading message when no playlist is passed in 1`] = `
<p>
  Loading
</p>
`;

exports[`PlaylistHeader should properly display playlist information when a playlist is passed in 1`] = `
<div
  className="row"
  style={
    Object {
      "marginBottom": 15,
    }
  }
>
  <div
    className="col-auto"
  >
    <img
      className=""
      height={250}
      src="https://spotify.com"
      style={
        Object {
          "border": "1px solid black",
        }
      }
      width={250}
    />
  </div>
  <div
    className="col"
  >
    <h2>
      My Playlist
    </h2>
    <p>
      My description
    </p>
    <p>
      Owned by: 
      My user
    </p>
    <p>
      0
       Songs. 
    </p>
    <div
      className="dropdown btn-group"
      onKeyDown={[Function]}
      role="group"
    >
      <button
        className="btn btn-info"
        disabled={true}
        type="button"
      >
        Play
      </button>
      <button
        aria-expanded={false}
        aria-haspopup={true}
        className="dropdown-toggle dropdown-toggle-split btn btn-info"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.247 11.14L2.451 5.658C1.885 5.013 2.345 4 3.204 4h9.592a1 1 0 0 1 .753 1.659l-4.796 5.48a1 1 0 0 1-1.506 0z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
